addon/object-diff.js

```diff
@@ -0,0 +1,184 @@
+import _ from 'lodash'
+
+export const ARRAY_PLACEHOLDER = '<%ARRAY_PLACEHOLDER%> elements equal'
+
+function diffValue (obj1Value, obj2Value) {
+  if (_.isArray(obj1Value) && _.isArray(obj2Value)) {
+    return diffArray(obj1Value, obj2Value)
+  } else if (_.isObject(obj1Value) && _.isObject(obj2Value)) {
+    return diffObject(obj1Value, obj2Value)
+  } else if (!_.isEqual(obj1Value, obj2Value)) {
+    return {
+      differentKeys: [undefined],
+      leftOnlyKeys: [],
+      rightOnlyKeys: [],
+      leftDiff: obj1Value,
+      rightDiff: obj2Value
+    }
+  }
+  return {
+    differentKeys: [],
+    leftOnlyKeys: [],
+    rightOnlyKeys: []
+  }
+}
+function completePath (key, nextKey) {
+  if (nextKey === undefined) {
+    return key
+  }
+  return `${key}.${nextKey}`
+}
+function diffArray (obj1Value, obj2Value) {
+  const leftDiff = []
+  const rightDiff = []
+  const differentKeys = []
+  const rightOnlyKeys = []
+  const leftOnlyKeys = []
+
+  let length = Math.min(obj1Value.length, obj2Value.length)
+  let i
+  for (i = 0; i < length; ++i) {
+    const result = diffValue(obj1Value[i], obj2Value[i])
+    const completeIndexPath = _.partial(completePath, i)
+    let hasMinusDiff = false
+    let hasRightDiff = false
+
+    if (result.leftOnlyKeys.length > 0) {
+      leftOnlyKeys.push.apply(leftOnlyKeys, _.map(result.leftOnlyKeys, completeIndexPath))
```

Comment:
It's a little harder to grep. Since it's done so many times in this file, maybe you can create a utility called `fast-concat` or something that modifies the array in-place.